_posts/2021-12-17-glean-in-2021.md

```diff
@@ -0,0 +1,188 @@
+and users can query the data in the columns where they expect them.
+
+[rate]: https://mozilla.github.io/glean/book/reference/metrics/rate.html
+[url]: https://mozilla.github.io/glean/book/reference/metrics/url.html
+[text]: https://mozilla.github.io/glean/book/reference/metrics/text.html
+
+### Revamped testing APIs
+
+We implemented some of the new testing APIs, namely [metric coverage support][coverage-support] and [`Ping.testBeforeNextSubmit`][beforeNextSubmit].
+Glean and FOG already use these new testing capabilities, but to my knowledge other Glean users are not yet relying on that.
+Certainly an area we should make more progress on in the coming year.
+
+[coverage-support]: https://github.com/mozilla/glean/pull/1482
+[beforeNextSubmit]: https://github.com/mozilla/glean/pull/1507
+
+### UniFFI - generate all the code
+
+Migration to rely on [UniFFI] for code generation has started!
+The [`uniffi` branch][uniffi-branch] is used for slowly migrating Glean piece by piece.
+Most of the metric types are [defined in UDL now][glean.udl] and the Kotlin code base has been migrated to it already,
+with all tests passing again.
+Work on this will continue next year, migrating the Swift and Python SDKs and then eventually merging it back into the main branch,
+so we can ship it to users.
+From there adding new metrics will require less implementation work and thus will allow us to follow up with some of the existing but currently paused requests.
+
+[uniffi]: https://github.com/mozilla/uniffi-rs/
+[uniffi-branch]: https://github.com/mozilla/glean/tree/uniffi
+[glean.udl]: https://github.com/badboy/glean/blob/4ae34b0f217160924a8c74165e387e190937fd33/glean-core/src/glean.udl
+
+## All the other things
+
+### Glean
+
+We [released][release-page] 9 major versions of Glean and did 33 releases in total.
+We're very liberal with breaking changes & major releases,
+though we make sure that upgrades require minimal changes where possible.
+The Kotlin, Swift, Python & Rust SDKs are versioned all the same,
+so a breaking change in only one SDK will still cause a major release for other SDKs, even without breaking changes.
+
+Some more products and projects are now using Glean, including [Rally studies][rally] and the [Mozilla VPN client][vpn].
+A full list of products is available in the [Glean Dictionary][dictionary].
+
+[release-page]: https://github.com/mozilla/glean/releases
+[rally]: https://rally.mozilla.org/
+[vpn]: https://www.mozilla.org/products/vpn/
+[dictionary]: https://dictionary.telemetry.mozilla.org/
+
+To ease integration between telemetry collected in Gecko and telemetry collected in the Android parts of the browser
+we're now [shipping Glean through GeckoView][geckoview], [we crashed it][crash], fixed it and shipped it again without new bugs cropping up.
+There's some work to be done next year to make local development easier again and to automate some of the tasks of upgading Glean across applications.
```

Comment:
The comma splices bug me a bit, and I like parentheses, but I'm not sure this is necessary.

```suggestion
To ease integration between telemetry collected in Gecko and telemetry collected in the Android parts of the browser
we're now [shipping Glean through GeckoView][geckoview]. (After [crashing it][crash], fixing it and shipping it again without new bugs cropping up.)
There's some work to be done next year to make local development easier again and to automate some of the tasks of upgading Glean across applications.
```